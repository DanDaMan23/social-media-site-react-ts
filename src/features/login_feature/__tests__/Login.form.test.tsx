import { render, screen, waitFor, fireEvent } from "@testing-library/react"
import { BrowserRouter } from "react-router-dom"
import { AuthenticationContext } from "../contexts/authentication.context"
import LoginForm from "../Login.form"

const renderLoginForm = (loginHandler: jest.Mock = jest.fn()) =>
  render(
    <BrowserRouter>
      <AuthenticationContext.Provider
        value={{ token: "", login: loginHandler, error: "" }}
      >
        <LoginForm />
      </AuthenticationContext.Provider>
    </BrowserRouter>
  )

describe("Login.form test", () => {
  it("Login form renders properly", () => {
    const { container } = renderLoginForm()

    expect(container).toMatchSnapshot()
  })

  it("Submits the form when it is valid", async () => {
    const mockLoginHandler = jest.fn()

    renderLoginForm(mockLoginHandler)

    const usernameInput = screen.getByRole("textbox", { name: /username/i })
    const passwordInput = screen.getByLabelText(/password/i)
    const submitButton = screen.getByRole("button", { name: /submit/i })

    fireEvent.input(usernameInput, { target: { value: "mockUsername" } })
    fireEvent.input(passwordInput, { target: { value: "mockPassword" } })
    fireEvent.submit(submitButton)

    await waitFor(() => {
      expect(mockLoginHandler).toHaveBeenCalledWith(
        "mockUsername",
        "mockPassword"
      )
    })
  })

  it("Fails to submit when form is invalid", async () => {
    const mockLoginHandler = jest.fn()

    renderLoginForm(mockLoginHandler)

    const submitButton = screen.getByRole("button", { name: /submit/i })

    fireEvent.submit(submitButton)

    await waitFor(() => {
      expect(mockLoginHandler).not.toHaveBeenCalled()
    })
  })
})
