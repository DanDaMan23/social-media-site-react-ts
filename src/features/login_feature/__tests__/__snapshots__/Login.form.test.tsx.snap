// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login.form test Login form renders properly 1`] = `
<div>
  <form
    class=""
  >
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="formBasicEmail"
      >
        Username
      </label>
      <input
        class="form-control"
        id="formBasicEmail"
        name="username"
        placeholder="Enter email"
        type="text"
      />
    </div>
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="formBasicPassword"
      >
        Password
      </label>
      <input
        class="form-control"
        id="formBasicPassword"
        name="password"
        placeholder="Password"
        type="password"
      />
    </div>
    <button
      class="btn btn-primary"
      disabled=""
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;
